storage-proofs/porep/src/nse/vanilla/nse.rs

```diff
@@ -399,19 +412,25 @@ impl<Tree: MerkleTreeTrait, G: Hasher> Clone for NodeProof<Tree, G> {
     }
 }
 
-/// Calculate the comm_r hash.
-pub fn hash_comm_r<D: Domain>(comm_layers: &[D], comm_replica: D) -> Fr {
-    assert!(comm_layers.len() <= MAX_COMM_R_ARITY, "too many layers");
-
-    let mut data: Vec<Fr> = Vec::with_capacity(comm_layers.len() + 1);
-    data.extend(comm_layers.iter().map(|v| {
-        let fr: Fr = (*v).into();
-        fr
-    }));
-    let comm_replica_fr: Fr = comm_replica.into();
-    data.push(comm_replica_fr);
-
-    let c = POSEIDON_CONSTANTS_15_BASE.with_length(data.len());
-    let mut p = Poseidon::new_with_preimage(&data, &c);
-    p.hash().into()
+// Hash all but the last layer's root to produce comm_layers.
+pub fn hash_comm_layers<D: Domain>(layer_roots_sans_last: &[D]) -> Fr {
+    let len = layer_roots_sans_last.len();
+    assert!(
+        len > 0 && len <= 15,
+        "layer roots exceed chosen Poseidon constants"
+    );
+    let consts = POSEIDON_CONSTANTS_15_BASE.with_length(len);
+    let layer_roots_sans_last: Vec<Fr> = layer_roots_sans_last
+        .iter()
+        .map(|&layer_root| layer_root.into())
+        .collect();
+    let mut hasher = Poseidon::new_with_preimage(&layer_roots_sans_last, &consts);
+    hasher.hash()
+}
+
+// Hashes comm_layers with the last layer's root (root_r).
+pub fn hash_comm_r<D: Domain>(comm_layers: D, root_r: D) -> Fr {
+    let preimg: [Fr; 2] = [comm_layers.into(), root_r.into()];
+    let mut hasher = Poseidon::new_with_preimage(&preimg[..], &*POSEIDON_CONSTANTS_2);
```

Comment:
This should probably use a newly-defined `POSEIDON_CONSTANTS_2_BASE` analogous to `POSEIDON_CONSTANTS_15_BASE`, rather than reuse the merkle-tree-specific `POSEIDON_CONSTANTS_2`.

We did reuse the merkle tree binary hash function for the equivalent purpose in SDR, but that was because domain separation tags for other function types had not yet been implemented.